torch/optim/adagrad.py

```diff
@@ -36,65 +36,55 @@ def __init__(self, params, lr=1e-2, lr_decay=0, weight_decay=0, initial_accumula
                 state['step'] = 0
-                state['sum'] = torch.full_like(p.data, initial_accumulator_value, memory_format=torch.preserve_format)
+                state['sum'] = torch.full_like(p.data, init_acc, memory_format=torch.preserve_format)
 
     def share_memory(self):
         for group in self.param_groups:
             for p in group['params']:
                 state = self.state[p]
                 state['sum'].share_memory_()
 
-    def step(self, closure=None):
-        """Performs a single optimization step.
-
-        Arguments:
-            closure (callable, optional): A closure that reevaluates the model
-                and returns the loss.
-        """
-        loss = None
-        if closure is not None:
-            loss = closure()
-
-        for group in self.param_groups:
-            for p in group['params']:
-                if p.grad is None:
-                    continue
-
-                grad = p.grad.data
-                state = self.state[p]
-
-                state['step'] += 1
-
-                if group['weight_decay'] != 0:
-                    if p.grad.data.is_sparse:
-                        raise RuntimeError("weight_decay option is not compatible with sparse gradients")
-                    grad = grad.add(group['weight_decay'], p.data)
-
-                clr = group['lr'] / (1 + (state['step'] - 1) * group['lr_decay'])
-
-                if grad.is_sparse:
-                    grad = grad.coalesce()  # the update is non-linear so indices must be unique
-                    grad_indices = grad._indices()
-                    grad_values = grad._values()
-                    size = grad.size()
-
-                    def make_sparse(values):
-                        constructor = grad.new
-                        if grad_indices.dim() == 0 or values.dim() == 0:
-                            return constructor().resize_as_(grad)
-                        return constructor(grad_indices, values, size)
-                    state['sum'].add_(make_sparse(grad_values.pow(2)))
-                    std = state['sum'].sparse_mask(grad)
-                    std_values = std._values().sqrt_().add_(group['eps'])
-                    p.data.add_(-clr, make_sparse(grad_values / std_values))
-                else:
-                    state['sum'].addcmul_(1, grad, grad)
-                    std = state['sum'].sqrt().add_(group['eps'])
-                    p.data.addcdiv_(-clr, grad, std)
-
-        return loss
+    def get_update(self, p, lr_decay=0, weight_decay=0, eps=1e-10, **_):
+        grad = p.grad.data
+        state = self.state[p]
+
+        if weight_decay > 0:
+            grad = grad.add(weight_decay, p.data)
+
+        c = 1 + state['step'] * lr_decay
+        state['step'] += 1
+        state['sum'].addcmul_(1, grad, grad)
+        std = state['sum'].sqrt().add_(eps)
+        return grad.div_(std) / c
+
+    def get_sparse_update(self, p, lr_decay=0, weight_decay=0, eps=1e-10, **_):
```

Comment:
This also raises an interesting point. If a sparse and non-sparse version of an algorithm don't convergence to the same point, should they be packaged under the same name (`get_update` vs `get_sparse_update`)? Should we in fact check that the two converges to nearby values, or even exact same? Is this part of the implicit contract when using this interface?

Thoughts?

Reply:
I didn't realise that the sparse version was only a rough approximation of Adam. This interface does imply that `sparse` and `dense` updates would be the same.

My Thoughts: 
1. I am not sure whether it is possible to implement Adam for sparse gradients so that it complies with the dense version, but I think that is the code that should go in `get_sparse_update` eventually, i.e. I agree with you.
2. Does it make sense to call this sparse algorithm "Adam" if it has completely different convergence properties? The Adam algorithm is already known to have issues with convergence on some edge cases (see the AMSgrad paper), but what if this sparse approximation has even more (severe) issues than the original paper? Also convergence speeds could differ significantly.

Probably, it makes sense to keep `SparseAdam` as it is for the refactoring, but I am not completely satisfied with this solution if I am honest.